Document "Web3.js" for asset hub smart contracts (#358)
polkadot-developers/polkadot-docs: develop/smart-contracts/evm-toolkit/libraries/web3-js.md

---
title: Web3.js
description: Learn how to interact with the Asset Hub chain using Web3.js, deploying Solidity contracts, and interacting with deployed smart contracts.
---

# Web3.js

## Introduction

> Interacting with blockchains typically requires an interface between your application and the network itself. [Web3.js](https://web3js.readthedocs.io/) offers this interface through a comprehensive collection of libraries, facilitating seamless interaction with the nodes using HTTP or WebSocket protocols. This guide illustrates how to utilize Web3.js specifically for interactions with the Asset Hub chain.

Reviewer: ```suggestion
Interacting with blockchains typically requires an interface between your application and the network itself. [Web3.js](https://web3js.readthedocs.io/){target=\_blank} offers this interface through a comprehensive collection of libraries, facilitating seamless interaction with the nodes using HTTP or WebSocket protocols. This guide illustrates how to utilize Web3.js specifically for interactions with the Asset Hub chain.
```

> Interacting with blockchains typically requires an interface between your application and the network itself. [Web3.js](https://web3js.readthedocs.io/){target=\_blank} offers this interface through a comprehensive collection of libraries, facilitating seamless interaction with the nodes using HTTP or WebSocket protocols. This guide illustrates how to utilize Web3.js specifically for interactions with the Asset Hub chain.

Reviewer 2: ```suggestion
Interacting with blockchains typically requires an interface between your application and the network. [Web3.js](https://web3js.readthedocs.io/){target=\_blank} offers this interface through a comprehensive collection of libraries, facilitating seamless interaction with the nodes using HTTP or WebSocket protocols. This guide illustrates how to utilize Web3.js specifically for interactions with the Asset Hub chain.
```

## Set Up the Project

To start working with Web3.js, begin by initializing your project:

```bash
npm init -y
```

## Install Dependencies

Next, install the Web3.js library:

```bash
npm install web3
```

## Set Up the Web3 Provider

> The foundation of any Web3.js application is the provider configuration. This establishes your connection to the Asset Hub network:

Reviewer 2: ```suggestion
The provider configuration is the foundation of any Web3.js application. The following example establishes a connection to the Asset Hub network. To use the example script, replace `INSERT_RPC_URL`, `INSERT_CHAIN_ID`, and `INSERT_CHAIN_NAME` with the appropriate values. For example, for Westend Asset Hub testnet, use these specific connection parameters:

    ```js
    const PROVIDER_RPC = {
        rpc: 'https://westend-asset-hub-eth-rpc.polkadot.io',
        chainId: 420420421,
        name: 'westend-asset-hub'
    };
    ```
```

Reviewer 2: May also need to unindent this code block here. Please take a look at what it looks like when you make the changes and adjust as needed. Please also check for this on the other code snippets. I think there are two more places we do this???

My goal is to tell them anything they need to do before running the script at the top. This information should be part of the instructions as they can't succeed unless they do it. Thank you!

Author: Okay, that's fine. This implies removing the note below, and I will add a small sentence between the PROVIDER_RPC definition and the `connectToProvider.js` script to create a "connection/link" between them

```javascript title="connectToProvider.js"
--8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/connectToProvider.js'
```

!!! note
    Replace `INSERT_RPC_URL`, `INSERT_CHAIN_ID`, and `INSERT_CHAIN_NAME` with the appropriate values. For example, For Westend Asset Hub testnet, use these specific connection parameters:

    ```js
    const PROVIDER_RPC = {
        rpc: 'https://westend-asset-hub-eth-rpc.polkadot.io',
        chainId: 420420421,
        name: 'westend-asset-hub'
    };
    ```

With the Web3 provider set up, you can start querying the blockchain.

For instance, to fetch the latest block number of the chain, you can use the following code snippet:

> ???- "fetchLastBlock.js"

Reviewer 2: ```suggestion
???- code "fetchLastBlock.js"
```

    ```javascript title="fetchLastBlock.js"
    --8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/fetchLastBlock.js'
    ```

## Compile Contracts

> Asset Hub requires contracts to be compiled to PolkaVM bytecode. This is achieved using the [`revive`](https://github.com/paritytech/revive){target=\_blank} compiler. Install the [`@parity/revive`](https://github.com/paritytech/js-revive){target=\_blank} library as a development dependency:

Reviewer: Link PolkaVM to a section explaining what it is

```bash
npm install --save-dev @parity/revive
```

> Here’s a simple storage contract that you can use to follow the process:

Reviewer 2: ```suggestion
Here's a simple storage contract that you can use to follow the process:
```

```solidity title="Storage.sol"
--8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/Storage.sol'
```

> With that, you can now create a `compile.js` snippet that transform your solidity code into PolkaVM bytecode:

Reviewer 2: ```suggestion
With that, you can now create a `compile.js` snippet that transforms your solidity code into PolkaVM bytecode:
```

```javascript title="compile.js"
--8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/compile.js'
```

> After compilation, you'll have the `.json` ABI file and `.polkavm` bytecode file ready for deployment.

Reviewer: Add a brief explanation about what each file is

## Contract Deployment

> To deploy your compiled contract to Asset Hub using Web3.js, you'll need an account with a private key to sign the deployment transaction. Here's how to deploy the contract:

Reviewer: I would add a sentence mentioning that the process is exactly the same as any EVM compatible chain for deployment and interaction

```javascript title="deploy.js"
--8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/deploy.js'
```

Replace `INSERT_RPC_URL`, `INSERT_PRIVATE_KEY`, and `INSERT_CONTRACT_NAME` with the appropriate values.

## Interact with the Contract

Once deployed, you can interact with your contract using Web3.js methods. Here's how to set a number and read it back:

```javascript title="updateStorage.js"
--8<-- 'code/develop/smart-contracts/evm-toolkit/libraries/web3-js/updateStorage.js'
```

Replace `INSERT_RPC_URL`, `INSERT_PRIVATE_KEY`, and `INSERT_CONTRACT_ADDRESS` with the appropriate values.

> ## Conclusion

Reviewer 2: ```suggestion
```

Reviewer 2: Let's remove the Conclusion heading as it's inconsistent with other pages. I suggest moving the "Replace..." information above the code snippet, leaving a blank line, and then having this final paragraph. 

Then, please Where to Go Next as the final heading after this paragraph. The one on this page is a nice example: /smart-contracts/evm-toolkit/libraries/ethers-js/#where-to-go-next

Those changes will make this page more consistent with the rest of the site. Thank you!

Author: okay, thanks! addressed in [4ba8eae](https://github.com/polkadot-developers/polkadot-docs/pull/358/commits/4ba8eae02e7531db7dd1531bb1c186e0d3d6353b)
